### PR TITLE
feat: Add YAML config file option

### PR DESCRIPTION
**Pull request by Leo10Gama**
#### Why is this change necessary?
With this addition, users will also be able to define their config files in YAML, adding additional support for those who may be unfamiliar with TOML.

#### How does it address the issue?
A YAML `FileManager` child class has been added, which handles reading and writing to and from YAML files. It is used by `SamConfig` to read and write in config options.

#### What side effects does this change have?
Config files in TOML will not be affected, but any YAML config files can be passed in and read as well.

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [x] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [x] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/lib/config/file_manager.py [hunk 1]
```diff
@@ -159,3 +161,90 @@ def put_comment(document: dict, comment: str) -> Any:
...
+        document = YamlFileManager._to_yaml(document)
+        document.yaml_set_start_comment(comment)
+        return document
+
+    @staticmethod
+    def _to_yaml(document: dict) -> Any:
+        """Ensure a dictionary-like object is a YAML document."""
+        stream = StringIO()
+        YamlFileManager.yaml.dump(document, stream)
+        return YamlFileManager.yaml.load(stream.getvalue())
```
Comment (lucashuy) [inline samcli/lib/config/file_manager.py:250]: Can we update the doc string here for the method?
Additionally, it looks like this opens an I/O stream to capture the string, can we close this stream after reading the value in, whether that be with a `.close()` or inside of a context?

Comment (Leo10Gama) [reply samcli/lib/config/file_manager.py:250]: How were you thinking to update the docstring? I checked through the `ruamel.yaml` docs and the return value seems to be `Any` for the parsed documents, and if it's set to something else, the linter complains about certain methods not being available.

Comment (lucashuy) [reply samcli/lib/config/file_manager.py:250]: Even though it is a pretty minimal method, the `document` parameter. As for the return value, though I'm not the biggest fan of `Any`, we could mention that the method returns a YAML document that is derived from an `dict` (or I guess more specifically, an `OrderedDict`)

**Review by lucashuy — APPROVED** [review]
Thanks for raising the PR

**Review by hnnasit — APPROVED** [review]
Looks good to me!

**Pull request merged**